src/main/java/com/google/cloud/dataproc/templates/dataplex/DataplexGCStoBQ.java

```diff
@@ -0,0 +1,426 @@
+   * @throws IOException when Google Credential setup fails
+   */
+  public Dataset<Row> applyCustomSql(Dataset<Row> newPartitionsDf) throws IOException {
+    if (customSqlGCSPath != null) {
+      LOGGER.info("Reading custom SQL from GCS path: {}", customSqlGCSPath);
+      String[] pathAsList =
+          customSqlGCSPath.replace(GCS_PATH_PREFIX, EMPTY_STRING).split(FORWARD_SLASH);
+      String BUCKET_NAME = pathAsList[0];
+      String OBJECT_NAME = Stream.of(pathAsList).skip(1).collect(Collectors.joining(FORWARD_SLASH));
+
+      StorageOptions options =
+          StorageOptions.newBuilder()
+              .setProjectId(projectId)
+              .setCredentials(GoogleCredentials.getApplicationDefault())
+              .build();
+
+      Storage storage = options.getService();
+      Blob blob = storage.get(BUCKET_NAME, OBJECT_NAME);
+      String custom_sql = new String(blob.getContent());
+      newPartitionsDf.createOrReplaceTempView(SPARK_SQL_OUTPUT_TABLE_TEMP_VIEW_NAME);
+      return spark.sql(custom_sql);
+    } else {
+      return newPartitionsDf;
+    }
+  }
+
+  /**
+   * Loads dataset to BigQuery
+   *
+   * @param newPartitionsDf dataset with new partitions
+   */
+  public void writeToBQ(Dataset<Row> newPartitionsDf) {
+    if (newPartitionsDf != null) {
+      LOGGER.info("Writing to target table: {}", targetTable);
+      DataFrameWriter dfWriter =
+          newPartitionsDf
+              .write()
+              .format(GCS_BQ_OUTPUT_FORMAT)
+              .option(GCS_BQ_OUTPUT, targetTable)
+              .option(GCS_BQ_TEMP_BUCKET, bqTempBucket)
+              .option(INTERMEDIATE_FORMAT_OPTION_NAME, INTERMEDIATE_FORMAT_ORC)
+              .option(
+                  DATAPLEX_GCS_BQ_CREATE_DISPOSITION_PROP_NAME,
+                  DATAPLEX_GCS_BQ_CREATE_DISPOSITION_CREATE_IF_NEEDED)
+              .mode(sparkSaveMode);
+
+      if (partitionField != null) {
+        dfWriter.option(DATAPLEX_GCS_BQ_PARTITION_FIELD_PROP_NAME, partitionField);
+      }
+      if (partitionType != null) {
+        dfWriter.option(DATAPLEX_GCS_BQ_PARTITION_TYPE_PROP_NAME, partitionType);
+      }
+
+      dfWriter.save();
+    }
+  }
+
+  public void runTemplate() {
+    try {
+      this.spark = SparkSession.builder().appName("Dataplex GCS to BQ").getOrCreate();
+      this.sqlContext = new SQLContext(spark);
+      checkInput();
+
+      for (int i = 0; i < entityList.size(); i += 1) {
+        this.entity = entityList.get(i);
+        this.entityBasePath = DataplexUtil.getBasePathEntityData(entity);
+        this.inputFileFormat = DataplexUtil.getInputFileFormat(entity);
+        this.targetTableName = entity.split(FORWARD_SLASH)[entity.split(FORWARD_SLASH).length - 1];
+        this.targetTable =
+            String.format(BQ_TABLE_NAME_FORMAT, projectId, targetDataset, targetTableName);
+        LOGGER.info("Processing entity: {}", entity);
+
+        List<String> partitionKeysList = DataplexUtil.getPartitionKeyList(entity);
```

Comment:
getPartitionKeyList() again calls Entity API which can avoided

Reply:
That's a good point. Entity schema is fetched more than once for one template execution. I think we could live with that in this first version and refactor later. Redundancy is not too high nor expensive and since the template is meant to be executed in batches (e.g. daily), we are far away from reaching a quota limit or impacting template performance.